File: python/dp/HouseRobber.py
'''
You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.

Input: nums = [1,2,3,1]
Output: 4
Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3).
Total amount you can rob = 1 + 3 = 4.

Input: nums = [2,7,9,3,1]
Output: 12
Explanation: Rob house 1 (money = 2), rob house 3 (money = 9) and rob house 5 (money = 1).
Total amount you can rob = 2 + 9 + 1 = 12.
'''
from typing import List

class HouseRobber:
    def rob(self, nums: List[int]) -> int:
        rob1, rob2 = 0, 0

        # [ rob1, rob2, n, n+1, ...]
        for n in nums:
            # Calculate the max loot if we rob the current house 'n'
            # We can either:
            # 1. Rob current house 'n' + loot from 'rob1' (house before previous)
            # 2. Skip current house and keep loot from 'rob2' (previous house)
            temp = max(n + rob1, rob2)
            rob1 = rob2
            rob2 = temp
            
        return rob2

if __name__ == "__main__":
    solution = HouseRobber()
    
    # Test Case 1
    nums1 = [1, 2, 3, 1]
    print(f"Max loot for {nums1}: {solution.rob(nums1)}") # Expected: 4

    # Test Case 2
    nums2 = [2, 7, 9, 3, 1]
    print(f"Max loot for {nums2}: {solution.rob(nums2)}") # Expected: 12
    
    # Test Case 3
    nums3 = [2, 1, 1, 2]
    print(f"Max loot for {nums3}: {solution.rob(nums3)}") # Expected: 4

File: src/main/java/com/naresh/a_dsalgo/dp/dp1/HouseRobber.java
package com.naresh.a_dsalgo.dp.dp1;

/**
 * Problem: House Robber
 * Description: You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed.
 * Constraint: You cannot rob two adjacent houses.
 */
public class HouseRobber {
    /**
     * Algorithm: Iterative DP with two variables. For each house, decide whether to rob it (nums[i] + rob_i-2) or skip it (rob_i-1).
     */
    public int rob(int[] nums) {
        // Pattern: DP (Space Optimized) | Time: O(n), Space: O(1)
        int rob1 = 0, rob2 = 0; // Max robbed up to previous two houses
        for (int n : nums) {
            var temp = Math.max(n + rob1, rob2); // Max of robbing current or skipping
            rob1 = rob2; // Update rob1 to previous rob2
            rob2 = temp; // Update rob2 to new max
        }
        return rob2; // Max amount robbed
    }
    // FAANG Tip: Space optimization from O(n) to O(1) is a common follow-up. Highlight the recurrence relation: f(i) = max(f(i-1), f(i-2) + nums[i]).

    public static void main(String[] args) {
        var nums = new int[]{1, 2, 3, 1};
        var result = new HouseRobber().rob(nums);
        System.out.println("Max robbed: " + result);
    }

    /**
     * Dry Run:
     * Input: nums = {1, 2, 3, 1}
     *
     * 1. Initialization:
     *    rob1 = 0, rob2 = 0
     *
     * 2. Loop Iterations:
     *    n = 1: temp = max(1 + 0, 0) = 1, rob1 = 0, rob2 = 1
     *    n = 2: temp = max(2 + 0, 1) = 2, rob1 = 1, rob2 = 2
     *    n = 3: temp = max(3 + 1, 2) = 4, rob1 = 2, rob2 = 4
     *    n = 1: temp = max(1 + 2, 4) = 4, rob1 = 4, rob2 = 4
     *
     * 3. Result: 4
     */
}
